File: CPP/Animal.cpp
#include <iostream>
#include "a/Color.cpp"
#include "a/Size.cpp"

using namespace std;

string toString(Color c)
{
    switch (c)
    {
    case Color::BLACK:
        return "BLACK";
    case Color::WHITE:
        return "WHITE";
    case Color::BROWN:
        return "BROWN";
    case Color::DARKBROWN:
        return "DARKBROWN";
    case Color::GREY:
        return "GREY";
    case Color::BLACKSTRIPES:
        return "BLACKSTRIPES";
    case Color::WHITESTRIPES:
        return "WHITESTRIPES";
    }
    return "";
}

string toString(Size s)
{
    switch (s)
    {
    case Size::SMALL:
        return "SMALL";
    case Size::MEDIUM:
        return "MEDIUM";
    case Size::LARGE:
        return "LARGE";
    }
    return "";
}

class Animal
{
protected:
    Color color;
    Size size;

public:
    Animal(Color color, Size size) : color(color), size(size) {}

    virtual string sound() = 0;

    virtual string eat() = 0;
};

class Mammalia : public Animal
{
protected:
    int numberOfBabies;

public:
    Mammalia(Color color, Size size, int numberOfBabies) : Animal(color, size), numberOfBabies(numberOfBabies) {}

    virtual string run() = 0;
};

class Dog : public Mammalia
{
protected:
    bool fierce;

public:
    Dog(Color color, Size size, int numberOfBabies, bool fierce) : Mammalia(color, size, numberOfBabies), fierce(fierce) {}

    virtual string bark() = 0;

    virtual string bite() = 0;
};

class ThaiRidgeBack : public Dog
{
protected:
    string origin;

public:
    ThaiRidgeBack(Color color, Size size, int numberOfBabies, bool fierce, string origin) : Dog(color, size, numberOfBabies, fierce), origin(origin) {}

    string sound() override
    {
        return "Woof!!!";
    }

    string eat() override
    {
        return "Pet food";
    }

    string run() override
    {
        return "Run on the ground";
    }

    string bark() override
    {
        return "Bark loudly";
    }

    string bite() override
    {
        return "Vicious Bite";
    }

    void info()
    {
        cout << "ThaiRidgeBack: [ Color " << toString(color) << ", Size: " << toString(size) 
        << ", Sound: " << sound() << ", Eat: " << eat() << ", Run: " << run() << ", Bark: " << bark() 
        << ", Bite: " << bite() << ", NumOfBabies: " << numberOfBabies << ", Origin: " << origin << " ]" << endl;  
    }
};

class Aves : public Animal {
    public:
        Aves(Color color, Size size): Animal(color, size){}

        virtual string fly() = 0; 
};

class Bird : public Aves {
    protected:
        string egg;

    public:
        Bird(Color color, Size size, string egg): Aves(color, size), egg(egg){}

};

class HummingBird : public Bird {
    public:
        HummingBird(Color color, Size size, string egg): Bird(color, size,egg){}

        string sound() override {
            return "Chrip Chrip";
        }

        string eat() override {
            return "Seed";
        }

        string fly() override {
            return "Fly in the sky";
        }

        void info() {
            cout << "Humming Bird: [ Color: " << toString(color) << ", Size: " << toString(size) 
            << ", Sound: " << sound() << ", Eat: " << eat() << ", FLy: " << fly() << ", Egg: " << egg << " ]" << endl;
        }
};

class Osteicthyes : public Animal {
    public:
        Osteicthyes(Color color, Size size): Animal(color, size){}

        virtual string swimming() = 0; 
};

class Fish : public Osteicthyes {
    protected:
        string waterGroup;

    public:
        Fish(Color color, Size size, string waterGroup): Osteicthyes(color, size), waterGroup(waterGroup){}

};

class AngleFish : public Fish {
    protected:
        string location;

    public:
        AngleFish(Color color, Size size, string waterGroup, string location): Fish(color, size, waterGroup), location(location){}

        string sound() override {
            return "...";
        }

        string eat() override {
            return "Plankton";
        }

        string swimming() override {
            return "Swimming in the sea";
        }

        void info() {
            cout << "Angle Fish: [ Color: " << toString(color) << ", Size: " << toString(size) 
            << ", Sound: " << sound() << ", Eat: " << eat() << ", Swimming: " << swimming() 
            << "Water Group: " << waterGroup << ", Location: " << location << " ]" << endl; 
        }
};

File: Java/java_oopwork09/src/Animal.java
import a.*;

public abstract class Animal {
    protected Color color;
    protected Size size;

    public Animal(Color color, Size size) {
        this.color = color;
        this.size = size;
    }

    abstract String sound();

    abstract String eat();
}

abstract class Manmalia extends Animal {
    protected int numberofBabies;

    public Manmalia(Color color, Size size, int numberofBabies) {
        super(color, size);
        this.numberofBabies = numberofBabies;
    }

    abstract String run();
}

abstract class Dog extends Manmalia {
    protected boolean fierce;

    public Dog(Color color, Size size, int numberofBabies, boolean fierce) {
        super(color, size, numberofBabies);
        this.fierce = fierce;
    }

    abstract String bark();

    abstract String bite();
}

class ThaiRidgeBack extends Dog {
    protected String origin;

    public ThaiRidgeBack(Color color, Size size, int numberofBabies, boolean fierce, String origin) {
        super(color, size, numberofBabies, fierce);
        this.origin = origin;
    }

    @Override
    public String sound() {
        return "Woof";
    }

    @Override
    public String eat() {
        return "Meat";
    }

    @Override
    public String run() {
        return "Running";
    }

    @Override
    public String bark() {
        return "Bark";
    }

    @Override
    public String bite() {
        return "Bite";
    }

    public void info() {
        System.out.println("ThaiRidgeBack: [ Color: " + color + ", Size: " + size + ", Sound: " + sound() + ", Eat: "
                + eat() + ", Run: " + run() + ", Bark: " + bark() + ", Bite: " + bite() + ", Fierce: " + fierce + " ]");
    }
}

abstract class Aves extends Animal {

    public Aves(Color color, Size size) {
        super(color, size);
    }

    abstract String fly();
}

abstract class Bird extends Aves {
    protected String egg;

    public Bird(Color color, Size size, String egg) {
        super(color, size);
        this.egg = egg;
    }

}

class HummingBird extends Bird {

    public HummingBird(Color color, Size size, String egg) {
        super(color, size, egg);
    }

    @Override
    public String sound() {
        return "Chrip";
    }

    @Override
    public String eat() {
        return "Seed";
    }

    @Override
    public String fly() {
        return "Fly in the sky";
    }

    public void info() {
        System.out.println("HummingBird: [ Color: " + color + ", Size: " + size + ", Sound: " + sound() + ", Eat: "
                + eat() + ", Fly: " + fly() + " ]");
    }

}

abstract class Osteicthyes extends Animal {

    public Osteicthyes(Color color, Size size) {
        super(color, size);
    }

    abstract String swimming();

}

abstract class Fish extends Osteicthyes {
    protected String waterGroup;

    public Fish(Color color, Size size, String waterGroup) {
        super(color, size);
        this.waterGroup = waterGroup;
    }

}

class AngleFish extends Fish {
    protected String location;

    public AngleFish(Color color, Size size, String waterGroup, String location) {
        super(color, size, waterGroup);
        this.location = location;
    }

    @Override
    public String sound() {
        return "...";
    }

    @Override
    public String eat() {
        return "Plankton";
    }

    @Override
    public String swimming() {
        return "Swimming in the sea";
    }

    public void info() {
        System.out.println("AngleFish: [ Color: " + color + ", Size: " + size + ", Sound: " + sound() + ", Eat: "
                + eat() + ", Swimming: " + swimming() + ", Water Group: " + waterGroup + ", Location: " + location
                + " ]");
    }
}